extra_data/components.py

```diff
@@ -425,6 +426,112 @@ def write_virtual_cxi(self, filename):
         VirtualCXIWriter(self).write(filename)
 
+    def write_frames(self, filename, trains, pulses):
+        """Write selected detector frames to a new EuXFEL HDF5 file
+
+        trains and pulses should be 1D arrays of the same length, containing
+        train IDs and pulse IDs (corresponding to the pulse IDs recorded by
+        the detector). i.e. (trains[i], pulses[i]) identifies one frame.
+        """
+        if (trains.ndim != 1) or (pulses.ndim != 1):
+            raise ValueError("trains & pulses must be 1D arrays")
+        inc_tp_ids = zip_trains_pulses(trains, pulses)
+
+        writer = FramesFileWriter(filename, self.data, inc_tp_ids)
+        try:
+            writer.write()
+        finally:
+            writer.file.close()
+
+
+def zip_trains_pulses(trains, pulses):
+    """Combine two similar arrays of train & pulse IDs as one struct array
+    """
+    if trains.shape != pulses.shape:
+        raise ValueError(
+            f"Train & pulse arrays don't match ({trains.shape} != {pulses.shape})"
+        )
+
+    res = np.zeros(trains.shape, dtype=np.dtype([
+        ('trainId', np.uint64), ('pulseId', np.uint64)
+    ]))
+    res['trainId'] = trains
+    res['pulseId'] = pulses
+    return res
+
+
+class FramesFileWriter(FileWriter):
+    """Write selected detector frames in European XFEL HDF5 format"""
+    def __init__(self, path, data, inc_tp_ids):
+        super().__init__(path, data)
+        self.inc_tp_ids = inc_tp_ids
+
+    def prepare_source(self, source):
+        """Prepare all the datasets for one source.
+
+        We do this as a separate step so the contents of the file are defined
+        together before the main data.
+        """
+        for key in sorted(self.data.keys_for_source(source)):
+            path = 'INSTRUMENT/{}/{}'.format(source, key.replace('.', '/'))
+            if key.startswith('image.'):
+                nentries = 0  # dataset grows later
+            else:
+                nentries = self.data.get_data_counts(source, key).sum()
+            src_ds1 = self.data._source_index[source][0].file[path]
+            self.file.create_dataset_like(
+                path, src_ds1, shape=(nentries,) + src_ds1.shape[1:],
+            )
+            self.data_sources.add(f"INSTRUMENT/{source}/{key.partition('.')[0]}")
+
+    def copy_image_data(self, source, keys):
+        """Copy selected frames of the detector image data"""
+        frame_tids_piecewise = []
+
+        src_files = sorted(
+            self.data._source_index[source],
+            key=lambda fa: fa.train_ids[0]
+        )
+
+        for fa in src_files:
+            _, counts = fa.get_index(source, 'image')
+            file_tids = np.repeat(fa.train_ids, counts.astype(np.intp))
+            file_pids = fa.file[f'/INSTRUMENT/{source}/image/pulseId'][:]
+            if file_pids.ndim == 2 and file_pids.shape[1] == 1:
+                # Raw data has a spurious extra dimension
+                file_pids = file_pids[:, 0]
```

Comment:
`numpy.squeeze`?

Reply:
I prefer to use an explicit check where it's practical, because if a 'real' dimension happens to have length 1, plain `np.squeeze(x)` will discard that as well.

You can tell it to squeeze a specific axis, but in our case we'd need to ensure that axis exists first.